// deezer-node/module/getTracks.js
const fetch = require("axios").default;
const { url } = require("../config.json");
require("../prototypes/turkishToEnglish.js");
module.exports = async(tracksUrl) => {
    
    const res = await fetch(tracksUrl).catch(err => { return undefined; });
    if(!res) return;
    
    const data = await res?.data?.data;
    const tracks = [];    

    for(var i = 0; i < data.length; i++) {
        const artists = [];

        data[i]?.contributors?.forEach(art => { 
            artists.push({
                name: art?.name,
                id: art?.id,
                visit: art?.link,
                image: art?.picture,
                type: art?.type,
                role: art?.role,
                tracks: art?.tracklist 
            });
        });

    if(!data[i].contributors) {
        artists.push({
            name: data[i].artist?.name,
            id: data[i].artist?.id,
            visit: "http://www.deezer.com/artist/"+data[i].artist?.id,
            image: "http://api.deezer.com/artist/"+data[i].artist?.id+"/image",
            type: data[i].artist?.type,
            tracks: data[i].artist?.tracklist 
        });
    };

        const trackObject = {
            name: data[i]?.title,
            id: data[i]?.id,
            visit: data[i]?.link,
            duration: (data[i]?.duration * 60),
            list: data[i]?.rank,
            preview: data[i]?.preview,
            explicitLyrics: data[i]?.explicit_lyrics,
            artists: artists,
            type: data[i]?.type
        };

        await tracks.push(trackObject);
    };
    
    return tracks;
}
